src/Streamly/Coreutils/Cp.hs

```diff
@@ -0,0 +1,258 @@
+module Streamly.Coreutils.Cp
+    ( CpOptions (..)
+    , defaultCpOptions
+    , copy
+    , traverseDir
+    , cpFile
+    , cpFiles
+    , cpDirSerial
+    , cpDirAsync
+    , cpDirParallel
+   )
+where
+
+import Path
+import Streamly
+
+import qualified Streamly.Prelude as S
+import qualified Streamly.Internal.Prelude as IP
+import qualified Streamly.Internal.FileSystem.File as File
+
+import System.IO.Unsafe (unsafePerformIO)
+import System.Directory (listDirectory, doesFileExist, createDirectoryIfMissing)
+
+-------------------------------------------------------------------------------
+-- Record for options used with cp
+-------------------------------------------------------------------------------
+
+-- | Data type to represent flags\/options in GNU @cp@
+--
+-- @since 0.1.0.0
+data CpOptions = CpOptions
+    { verbose :: Bool
+    -- ^ Prints the source and destination files being copied
+    , force :: Bool
+    -- ^ if the destination file
+    -- cannot be opened, removes it and continues copying the rest
+    , overwriteExisting :: Bool
+    -- ^ If @True@, an existing destination file will be overwritten
+    -- due to copying
+    }
+
+
+-- | Default options for @cp@ are @verbose@ and @overwriteExisting@ to
+-- @True@ and @force@ set to @False@
+--
+-- @since 0.1.0.0
+{-# INLINE defaultCpOptions #-}
+defaultCpOptions :: CpOptions
+defaultCpOptions = CpOptions True False True
+
+
+-------------------------------------------------------------------------------
+-- cp and helper functions for options
+-------------------------------------------------------------------------------
+
+
+-- | Prints paths of source, destination files when @verbose@ is @True@.
+-- Returns @True@ if copying operation was successful.
+-- @False@ is returned in cases where a file by the same name
+-- exists in the destination path and @overwriteExisting@ is set to
+-- @False@ or if the source file does not exist.
+--
+-- @since 0.1.0.0
+{-# INLINE copy #-}
+copy
+    :: CpOptions
+    -> Path Abs File
+    -> Path Abs File
+    -> IO Bool
+copy opt src dst = do
+    let srcFp = fromAbsFile src
+    let dstFp = fromAbsFile dst
+    existDst <- doesFileExist dstFp
+    existSrc <- doesFileExist srcFp
+    if not existSrc
+    then do
+        putStrLn $ "Source file " ++ srcFp ++ " does not exist"
+        return False
+    else if existDst && not (overwriteExisting opt)
+    then do
+        putStrLn $ "overwriteExisting : False " ++ dstFp ++ " already exists"
+        return False
+    else if verbose opt == True
+    then do
+        File.fromChunks dstFp $ File.toChunksWithBufferOf (256*1024) srcFp
+        putStrLn $ srcFp ++ " -> " ++ dstFp
+        return True
+    else do
+        File.fromChunks dstFp $ File.toChunksWithBufferOf (256*1024) srcFp
+        return True
```

Comment:
Doesn't look like you're handling the appropriate use case when `force` is true and you don't have the required permission.